### DISCUSSION TITLE
After migrating to Cake.Frosting v1, getting MSBuild errors

### DISCUSSION THREAD

**Discussion by bdukes** [cake-build/cake#2957] [Q&A] [answered]
We were looking to [move to using Cake.Frosting in DNN Platform](https://github.com/dnnsoftware/Dnn.Platform/pull/4322), and thought we fairly successfully did that migration.  However, after tweaking the few differences that cropped up, we ended up with an MSBuild error that we can't resolve when running in Azure Pipelines.  We've tried adding the UseDotNet task to make sure everything works and adjusting the framework the Frosting project is built with, but are getting these weird errors from a project targeting .NET Standard 2.0:

[link to log](https://dev.azure.com/dotnet/DNN/_build/results?buildId=42243&view=logs&j=12f1170f-54f2-53f3-20dd-22fc7dff55f9&t=d01ed771-bc1f-5e98-e7ad-71f8c6eef93d&l=4190)
>  D:\a\1\s\DNN Platform\DotNetNuke.Abstractions\DotNetNuke.Abstractions.csproj : error : Unable to locate the .NET SDK. Check that it is installed and that the version specified in global.json (if any) matches the installed version.

[link to log](https://dev.azure.com/dotnet/DNN/_build/results?buildId=42243&view=logs&j=12f1170f-54f2-53f3-20dd-22fc7dff55f9&t=d01ed771-bc1f-5e98-e7ad-71f8c6eef93d&l=4196)
>  D:\a\1\s\DNN Platform\DotNetNuke.Abstractions\DotNetNuke.Abstractions.csproj : error MSB4236: The SDK 'Microsoft.NET.Sdk' specified could not be found.

Has anyone run into these?  Any ideas of what might be missing in our Azure Pipelines environment?

cc: @valadas @donker @mitchelsellers

**Comment by pascalberger** [comment]
On first glance I don't see anything directly related to Cake which can cause this. You have this issue only while running on Azure Pipelines? Do you build also on other CI systems?

Some ideas how to analyze further:
* Can you set the `MSBuildSDKsPath` environment variable to the path where the SDK is to try if it works afterwards?
* You build with the correct MsBuild configuration?

**Reply by valadas** [reply]
We are building with this configuration:

https://github.com/dnnsoftware/Dnn.Platform/blob/8f63262869afc3f14b7053dea455e6439bedfd3d/Build/Cake/compiling.cs

**Reply by pascalberger** [reply]
Can you reproduce the behavior locally? I tried to run your pipeline local on my machine.

When running `.\build.ps1 --target=RestoreNuGetPackages` I get the following error message in the console (it will not fail the build though):

```
========================================
RestoreNuGetPackages
========================================
MSBuild auto-detection: using msbuild version '16.7.0.37604' from 'C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\Current\bin'.
C:\Git\Dnn.Platform\DNN Platform\DotNetNuke.ModulePipeline\DotNetNuke.ModulePipeline.csproj : error : Unable to locate the .NET Core SDK. Check that it is installed and that the version specified in global.json (if any) matches the installed version.
C:\Git\Dnn.Platform\DNN Platform\DotNetNuke.ModulePipeline\DotNetNuke.ModulePipeline.csproj : error MSB4236: The SDK 'Microsoft.NET.Sdk' specified could not be found.
C:\Git\Dnn.Platform\DNN Platform\DotNetNuke.DependencyInjection\DotNetNuke.DependencyInjection.csproj : error MSB4236: The SDK 'Microsoft.NET.Sdk' specified could not be found.
C:\Git\Dnn.Platform\Build\BuildScripts\AEModule.build(3,3): warning MSB4011: "C:\Git\Dnn.Platform\packages\Yarn.MSBuild.1.16.0\build\Yarn.MSBuild.props" cannot be imported again. It was already imported at "C:\Git\Dnn.Platform\Dnn.AdminExperience\Dnn.PersonaBar.Extensions\Dnn.PersonaBar.Extensions.csproj (3,3)". This is most likely a build authoring error. This subsequent import will be ignored.
C:\Git\Dnn.Platform\Dnn.AdminExperience\Dnn.PersonaBar.Extensions\Dnn.PersonaBar.Extensions.csproj(772,3): warning MSB4011: "C:\Git\Dnn.Platform\packages\Yarn.MSBuild.1.16.0\build\Yarn.MSBuild.targets" cannot be imported again. It was already imported at "C:\Git\Dnn.Platform\Build\BuildScripts\AEModule.build (4,3)". This is most likely a build authoring error. This subsequent import will be ignored.
C:\Git\Dnn.Platform\DNN Platform\DotNetNuke.Abstractions\DotNetNuke.Abstractions.csproj : error MSB4236: The SDK 'Microsoft.NET.Sdk' specified could not be found.
  It was not possible to find any installed .NET Core SDKs
  Did you mean to run .NET Core SDK commands? Install a .NET Core SDK from:
      https://aka.ms/dotnet-download

WARNING: Error reading msbuild project information, ensure that your input solution or project file is valid. NETCore and UAP projects will be skipped, only packages.config files will be restored.
All packages listed in packages.config are already installed.
```

Running with verbose diagnostic (`.\build.ps1 --target=RestoreNuGetPackages --verbosity=diagnostic`) adds the command which is called from the build script:

```
========================================
RestoreNuGetPackages
========================================
Executing task: RestoreNuGetPackages
Executing: "C:/Git/Dnn.Platform/tools/NuGet.CommandLine.5.8.0/tools/NuGet.exe" restore "C:/Git/Dnn.Platform/DNN_Platform.sln" -NonInteractive
MSBuild auto-detection: using msbuild version '16.7.0.37604' from 'C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\Current\bin'.
C:\Git\Dnn.Platform\DNN Platform\DotNetNuke.ModulePipeline\DotNetNuke.ModulePipeline.csproj : error : Unable to locate the .NET Core SDK. Check that it is installed and that the version specified in global.json (if any) matches the installed version.
C:\Git\Dnn.Platform\DNN Platform\DotNetNuke.ModulePipeline\DotNetNuke.ModulePipeline.csproj : error MSB4236: The SDK 'Microsoft.NET.Sdk' specified could not be found.
C:\Git\Dnn.Platform\DNN Platform\DotNetNuke.DependencyInjection\DotNetNuke.DependencyInjection.csproj : error MSB4236: The SDK 'Microsoft.NET.Sdk' specified could not be found.
C:\Git\Dnn.Platform\Build\BuildScripts\AEModule.build(3,3): warning MSB4011: "C:\Git\Dnn.Platform\packages\Yarn.MSBuild.1.16.0\build\Yarn.MSBuild.props" cannot be imported again. It was already imported at "C:\Git\Dnn.Platform\Dnn.AdminExperience\Dnn.PersonaBar.Extensions\Dnn.PersonaBar.Extensions.csproj (3,3)". This is most likely a build authoring error. This subsequent import will be ignored.
C:\Git\Dnn.Platform\Dnn.AdminExperience\Dnn.PersonaBar.Extensions\Dnn.PersonaBar.Extensions.csproj(772,3): warning MSB4011: "C:\Git\Dnn.Platform\packages\Yarn.MSBuild.1.16.0\build\Yarn.MSBuild.targets" cannot be imported again. It was already imported at "C:\Git\Dnn.Platform\Build\BuildScripts\AEModule.build (4,3)". This is most likely a build authoring error. This subsequent import will be ignored.
C:\Git\Dnn.Platform\DNN Platform\DotNetNuke.Abstractions\DotNetNuke.Abstractions.csproj : error MSB4236: The SDK 'Microsoft.NET.Sdk' specified could not be found.
  It was not possible to find any installed .NET Core SDKs
  Did you mean to run .NET Core SDK commands? Install a .NET Core SDK from:
      https://aka.ms/dotnet-download

WARNING: Error reading msbuild project information, ensure that your input solution or project file is valid. NETCore and UAP projects will be skipped, only packages.config files will be restored.
All packages listed in packages.config are already installed.
Finished executing task: RestoreNuGetPackages
```

If I now run the same command which is run from the build script locally I get the same error:

```
PS C:\Git\Dnn.Platform> C:/Git/Dnn.Platform/tools/NuGet.CommandLine.5.8.0/tools/NuGet.exe restore "C:/Git/Dnn.Platform/DNN_Platform.sln" -NonInteractive
MSBuild auto-detection: using msbuild version '16.7.0.37604' from 'C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\Current\bin'.
C:\Git\Dnn.Platform\DNN Platform\DotNetNuke.ModulePipeline\DotNetNuke.ModulePipeline.csproj : error : Unable to locate the .NET Core SDK. Check that it is installed and that the version specified in global.json (if any) matches the installed version.
C:\Git\Dnn.Platform\DNN Platform\DotNetNuke.ModulePipeline\DotNetNuke.ModulePipeline.csproj : error MSB4236: The SDK 'Microsoft.NET.Sdk' specified could not be found.
C:\Git\Dnn.Platform\DNN Platform\DotNetNuke.DependencyInjection\DotNetNuke.DependencyInjection.csproj : error MSB4236: The SDK 'Microsoft.NET.Sdk' specified could not be found.
C:\Git\Dnn.Platform\Build\BuildScripts\AEModule.build(3,3): warning MSB4011: "C:\Git\Dnn.Platform\packages\Yarn.MSBuild.1.16.0\build\Yarn.MSBuild.props" cannot be imported again. It was already imported at "C:\Git\Dnn.Platform\Dnn.AdminExperience\Dnn.PersonaBar.Extensions\Dnn.PersonaBar.Extensions.csproj (3,3)". This is most likely a build authoring error. This subsequent import will be ignored.
C:\Git\Dnn.Platform\Dnn.AdminExperience\Dnn.PersonaBar.Extensions\Dnn.PersonaBar.Extensions.csproj(772,3): warning MSB4011: "C:\Git\Dnn.Platform\packages\Yarn.MSBuild.1.16.0\build\Yarn.MSBuild.targets" cannot be imported again. It was already imported at "C:\Git\Dnn.Platform\Build\BuildScripts\AEModule.build (4,3)". This is most likely a build authoring error. This subsequent import will be ignored.
C:\Git\Dnn.Platform\DNN Platform\DotNetNuke.Abstractions\DotNetNuke.Abstractions.csproj : error MSB4236: The SDK 'Microsoft.NET.Sdk' specified could not be found.
  It was not possible to find any installed .NET Core SDKs
  Did you mean to run .NET Core SDK commands? Install a .NET Core SDK from:
      https://aka.ms/dotnet-download

WARNING: Error reading msbuild project information, ensure that your input solution or project file is valid. NETCore and UAP projects will be skipped, only packages.config files will be restored.
All packages listed in packages.config are already installed.
```

Is this the error you experience on your CI builds? If so this would keep Frosting out of the equation, since it seems to be an issue somewhere else (Cake in this case just calls the nuget.exe process, which leads to the error).

**Comment by bdukes** [comment]
Our build is working fine before using Cake.Frosting, and then using the
same build logic with the Frosting runner we see this error. The
configuration is the same in either case.

Our current theory is that when the runner downloads .Net 5 SDK during
bootstrapping it messes with the config for other steps (though that
doesn't seem to get happening locally). It's still pulling in an RC version
of the SDK, could that be an issue?

I've seen suggestions to set MSBuildSDKsPath, but most of them ended with
Microsoft folks saying to revert that change because they had fixed
whatever underlying issue. So, we would use the UseDotNet task to restore
to a certain location and then set that environment variable to the same
location?

We don't build on any other CI system.

Thanks!

**Reply by pascalberger** [reply]
You can use .NET 5 RTM for building. Frosting should work fine with it (we have an open issue for the RC version used in the template: https://github.com/cake-build/cake/issues/2956).

`MSBuildSDKsPath` should not be required, my suggestion was more to try it to find out what's exactly is going wrong. The way you suggested could be worth a try to see if it will build that way.

**Comment by valadas** [comment, marked as answer]
> You can use .NET 5 RTM for building

@pascalberger Can you explain in a bit more details what we need to adjust to do so ?

**Reply by pascalberger** [reply]
Change the entry in `build.config` to `DOTNET_VERSION=5.0.100`

**Reply by valadas** [reply]
Awesome, this appears to resolve the main issue, thanks a lot it was set to a specific RC2 version and this change allowed the build to continue further, actually everything builds fine but there is another probably unrelated issue running unit tests which I will look into shortly.

**Reply by pascalberger** [reply]
Glad you got it working.
